README.md
# [qhsdk](https://pypi.org/project/qhsdk/)

## [qhsdk](https://pypi.org/project/qhsdk/) 的介绍

[qhsdk](https://pypi.org/project/qhsdk/) 主要是为 **https://qhkch.com/** 提供 SDK 的 Python 库, 您可以通过 [奇货可查机构VIP接口文档](https://www.showdoc.cc/qhkchapi?page_id=3878436763477725) 了解和查询详细数据接口！

## [qhsdk](https://pypi.org/project/qhsdk/) 服务于 **www.qhkch.com**

<img src="https://static.qhkch.com/dist//style/res/logo.png" align = center/>

## [qhsdk](https://pypi.org/project/qhsdk/) 的特色

[qhsdk](https://pypi.org/project/qhsdk/) 主要改进如下:

1. [qhsdk](https://pypi.org/project/qhsdk/)支持 **Python 3.7** 及以上版本;
2. 目前提供已提供奇货可查全部接口;
3. 提供完善的接口文档, 提高 [qhsdk](https://pypi.org/project/qhsdk/) 的易用性;

# 安装方法

```
pip install qhsdk
```

# 升级方法

```
pip install qhsdk --upgrade
```

# 快速入门

目标数据: 奇货可查-商品-持仓数据接口

示例代码:

```python
import qhsdk as qh

pro = qh.pro_api(token="此处输入您的token, 请联系奇货可查获取！")

variety_positions_df = pro.variety_positions(fields="shorts", code="rb1810", date="2018-08-08")
print(variety_positions_df)
```

示例结果:

```
   broker  short  short_chge
0    银河期货  60987       -4228
1    永安期货  57520       -1071
2    中信期货  38120        -620
3    国泰君安  36498         528
4    方正中期  32105        4444
5    海通期货  29638       -2783
6    东海期货  29250         450
7    光大期货  28458         -84
8    南华期货  27853        -144
9    中辉期货  26101        -553
10   中大期货  23761        1572
11   鲁证期货  22501        -598
12   兴证期货  22262        -842
13   东证期货  21675        -686
14   徽商期货  18966        -607
15   中信建投  18583        -625
16   华泰期货  17076       -5797
17   国投安信  16808         349
18   申银万国  14876         376
19   广发期货  14588       -2196
20   大地期货      0      -14603
```

# 特别说明

## 致谢

特别感谢 [AkShare](https://github.com/jindaxiang/akshare) 项目提供借鉴学习的机会;

## 声明

1. [qhsdk](https://pypi.org/project/qhsdk) 提供的数据仅供参考, 不构成任何投资建议;
2. 任何基于 [qhsdk](https://pypi.org/project/qhsdk) 进行研究的投资者请注意数据风险;
3. [qhsdk](https://pypi.org/project/qhsdk) 的使用请遵循奇货可查网站的用户协议;
4. [qhsdk](https://pypi.org/project/qhsdk) 使用产生的所有问题的最终解释权归奇货可查网站所有;

# 版本更新说明
```
0.0.1
发布测试版本

0.0.2
调试接口

0.0.3
增加lxml=4.4.1

0.0.4
更新说明文档

0.0.5
新增通过用户名和密码登录并访问VIP资源功能

0.0.6
修正导入问题

0.0.7
更新 README 文档

0.0.8
第二版接口测试

0.0.9
更新说明文档

0.1.0
更新测试文件

0.1.1
预览版

0.1.2
增加 token 使用说明

0.1.3
增加 inventory 参数类型二

0.1.4
新增: 合约持仓数据接口, 商品持仓数据接口, 修复部分接口与文档一致

0.1.5
新增: 增加 broker_positions_process 中 start_date 和 end_date 可选参数

0.1.6
新增: 增加 variety_calc_positions 接口

0.1.7
修改: 修改 variety_all 接口, 新增 3 个字段

0.1.8
新增: 新增 variety_net_position_list 接口

0.1.9
新增: 同步文档 2021年11月25日 版本

0.1.10
新增: inventory_new 接口
修复：返回空值

0.1.11
修复: 修改 url 地址
```

# 文档构建

```shell
cd docs
```

```shell
pip install -r requirements.txt
```

```shell
sphinx-build -b html . _build
```


docs/changelog.md
# 更新记录

```
0.0.1
发布测试版本

0.0.2
调试接口

0.0.3
增加lxml=4.4.1

0.0.4
更新说明文档

0.0.5
新增通过用户名和密码登录并访问VIP资源功能

0.0.6
修正导入问题

0.0.7
更新 README 文档

0.0.8
第二版接口测试

0.0.9
更新说明文档

0.1.0
更新测试文件

0.1.1
预览版

0.1.2
增加 token 使用说明

0.1.3
增加 inventory 参数类型二

0.1.4
新增: 合约持仓数据接口, 商品持仓数据接口, 修复部分接口与文档一致

0.1.5
新增: 增加 broker_positions_process 中 start_date 和 end_date 可选参数

0.1.6
新增: 增加 variety_calc_positions 接口

0.1.7
修改: 修改 variety_all 接口, 新增 3 个字段

0.1.8
新增: 新增 variety_net_position_list 接口

0.1.9
新增: 同步文档 2021年11月25日 版本

0.1.10
新增: inventory_new 接口
修复：返回空值

0.1.11
修复: 修改 url 地址
```



qhsdk/qhkc/qhkc_index.py
# -*- coding:utf-8 -*-
# /usr/bin/env python
"""
Author: Albert King
date: 2019/9/30 13:58
contact: jindaxiang@163.com
desc: 奇货可查网站目前已经商业化运营, 特提供奇货可查-指数数据接口, 方便您程序化调用
注：期货价格为收盘价; 现货价格来自网络; 基差=现货价格-期货价格; 基差率=(现货价格-期货价格)/现货价格 * 100 %.
"""
from typing import AnyStr

import pandas as pd
import requests

from qhsdk.qhkc.cons import QHKC_INDEX_URL, QHKC_INDEX_TREND_URL, QHKC_INDEX_PROFIT_LOSS_URL


def qhkc_index(name: AnyStr = "奇货商品", url: AnyStr = QHKC_INDEX_URL):
    """
    奇货可查-指数-指数详情
    获得奇货可查的指数数据: '奇货黑链', '奇货商品', '奇货谷物', '奇货贵金属', '奇货饲料', '奇货软商品', '奇货化工', '奇货有色', '奇货股指', '奇货铁合金', '奇货油脂'
    :param url: 网址
    :param name 中文名称
    :return: pd.DataFrame
        date    price   volume  ...      margin     profit long_short_ratio
        2013-01-04     1000   260820  ...  1130485758    1816940            52.78
        2013-01-07  998.244   245112  ...  1132228518    2514410            52.15
        2013-01-08   1000.8   318866  ...  1160374489    2981010            51.99
        2013-01-09  998.661   247352  ...  1166611242    3904220            52.44
        2013-01-10  999.802   161292  ...  1153164771    1448190            52.81
               ...      ...      ...  ...         ...        ...              ...
        2019-09-24  845.391   881138  ...  1895149977  128379050             48.5
        2019-09-25  845.674   715180  ...  1797235248  128788230            48.29
        2019-09-26  840.154  1347570  ...  1730488227  137104890            48.44
        2019-09-27  834.831   920160  ...  1605342767  143128540            48.77
        2019-09-30  831.959  1031558  ...  1521875378  147810580            48.82
    """
    name_id_dict = {}
    qhkc_index_url = "https://qhkch.com/ajax/official_indexes.php"
    r = requests.post(qhkc_index_url)
    display_name = [item["name"] for item in r.json()["data"]]
    index_id = [item["id"] for item in r.json()["data"]]
    for item in range(len(display_name)):
        name_id_dict[display_name[item]] = index_id[item]
    payload_id = {
        "id": name_id_dict[name]
    }
    r = requests.post(url, data=payload_id)
    print(name, "数据获取成功")
    json_data = r.json()
    date = json_data["data"]["date"]
    price = json_data["data"]["price"]
    volume = json_data["data"]["volume"]
    open_interest = json_data["data"]["openint"]
    total_value = json_data["data"]["total_value"]
    profit = json_data["data"]["profit"]
    long_short_ratio = json_data["data"]["line"]
    df_temp = pd.DataFrame([date, price, volume, open_interest, total_value, profit, long_short_ratio]).T
    df_temp.columns = ["date", "price", "volume", "open_interest", "margin", "profit", "long_short_ratio"]
    return df_temp


def qhkc_index_trend(name: AnyStr = "奇货商品", url: AnyStr = QHKC_INDEX_TREND_URL):
    """
    奇货可查-指数-大资金动向
    获得奇货可查的指数数据: '奇货黑链', '奇货商品', '奇货谷物', '奇货贵金属', '奇货饲料', '奇货软商品', '奇货化工', '奇货有色', '奇货股指', '奇货铁合金', '奇货油脂'
    :param url: 网址
    :param name None
    :return: pd.DataFrame
        broker    grade    money     open_order     variety
        中金期货     B -3.68209e+07  3.68209e+07      沪金
        浙商期货     D    -25845534     25845534      沪银
        永安期货     A    -25614000     25614000      沪银
        招商期货     D    -23517351     23517351      沪银
        海通期货     A     21440845     21440845      沪金
         美尔雅     D     21370975     21370975      沪金
        中原期货     C    -21204612     21204612      沪银
        国投安信     A -1.52374e+07  1.52374e+07      沪银
        中信期货     C  1.50941e+07  1.50941e+07      沪银
        海通期货     A -1.47184e+07  1.47184e+07      沪银
        方正中期     E -1.31432e+07  1.31432e+07      沪银
        东证期货     D   -1.283e+07    1.283e+07      沪银
        一德期货     A  1.24973e+07  1.24973e+07      沪银
        国投安信     A    -11602860     11602860      沪金
        国泰君安     B -1.09363e+07  1.09363e+07      沪金
        华安期货     D -9.99499e+06  9.99499e+06      沪金
        南华期货     B -9.23675e+06  9.23675e+06      沪银
        国贸期货     B  8.55245e+06  8.55245e+06      沪银
        道通期货     C      8527675      8527675      沪金
        招商期货     D -7.85457e+06  7.85457e+06      沪金
        东方财富     E -7.58235e+06  7.58235e+06      沪银
        五矿经易     A  6.95354e+06  6.95354e+06      沪银
        银河期货     B  6.84522e+06  6.84522e+06      沪银
        国贸期货     B      6731025      6731025      沪金
        平安期货     D     -6710418      6710418      沪银
        上海中期     C      6628800      6628800      沪金
        中信期货     C     -6345830      6345830      沪金
        银河期货     B     -6126295      6126295      沪金
        华泰期货     A -5.96254e+06  5.96254e+06      沪金
        招金期货     E -5.53029e+06  5.53029e+06      沪银
        东证期货     D -5.47486e+06  5.47486e+06      沪金
        光大期货     C     -5334730      5334730      沪金
        广发期货     D  5.31904e+06  5.31904e+06      沪金
        国信期货     D -5.05211e+06  5.05211e+06      沪金
    """
    name_id_dict = {}
    qhkc_index_url = "https://qhkch.com/ajax/official_indexes.php"
    r = requests.post(qhkc_index_url)
    display_name = [item["name"] for item in r.json()["data"]]
    index_id = [item["id"] for item in r.json()["data"]]
    for item in range(len(display_name)):
        name_id_dict[display_name[item]] = index_id[item]
    payload_id = {
        "page": 1,
        "limit": 10,
        "index": name_id_dict[name],
        "date": ""
    }
    r = requests.post(url, data=payload_id)
    print(f"{name}期货指数-大资金动向数据获取成功")
    json_data = r.json()
    df_temp = pd.DataFrame()
    for item in json_data["data"]:
        broker = item["broker"]
        grade = item["grade"]
        money = item["money"]
        order_money = item["order_money"]
        variety = item["variety"]
        df_temp = df_temp.append(pd.DataFrame([broker, grade, money, order_money, variety]).T)
    df_temp.columns = ["broker", "grade", "money", "open_order", "variety"]
    df_temp.reset_index(drop=True, inplace=True)
    return df_temp


def qhkc_index_profit_loss(name: AnyStr = "奇货商品", url: AnyStr = QHKC_INDEX_PROFIT_LOSS_URL, start_date="",
                           end_date=""):
    """
    奇货可查-指数-盈亏详情
    获得奇货可查的指数数据: '奇货黑链', '奇货商品', '奇货谷物', '奇货贵金属', '奇货饲料', '奇货软商品', '奇货化工', '奇货有色', '奇货股指', '奇货铁合金', '奇货油脂'
    :param url: 网址
    :param name None
    :param start_date: ""
    :param end_date: "20190716" 指定 end_date 就可以了
    :return: pd.DataFrame
        indexes       value  trans_date
        招金期货-沪金  -307489200  2019-09-30
        平安期货-沪银  -195016650  2019-09-30
        建信期货-沪银  -160327350  2019-09-30
        国贸期货-沪银  -159820965  2019-09-30
        东证期货-沪银  -123508635  2019-09-30
            ...         ...         ...
        永安期货-沪银   187411350  2019-09-30
        中信期货-沪金   242699750  2019-09-30
        华泰期货-沪银   255766185  2019-09-30
        永安期货-沪金   293008700  2019-09-30
        国泰君安-沪金   302774950  2019-09-30
    """
    name_id_dict = {}
    qhkc_index_url = "https://qhkch.com/ajax/official_indexes.php"
    r = requests.post(qhkc_index_url)
    display_name = [item["name"] for item in r.json()["data"]]
    index_id = [item["id"] for item in r.json()["data"]]
    for item in range(len(display_name)):
        name_id_dict[display_name[item]] = index_id[item]
    payload_id = {
        "index": name_id_dict[name],
        "date1": start_date,
        "date2": end_date
    }
    r = requests.post(url, data=payload_id)
    print(f"{name}期货指数-盈亏分布数据获取成功")
    json_data = r.json()
    df_temp = pd.DataFrame()
    indexes = json_data["data"]["indexes"]
    value = json_data["data"]["value"]
    trans_date = [json_data["data"]["trans_date"]] * len(value)
    df_temp = pd.DataFrame([indexes, value, trans_date]).T
    df_temp.columns = ["indexes", "value", "trans_date"]
    return df_temp


if __name__ == "__main__":
    data = qhkc_index("奇货谷物")
    print(data)
    data = qhkc_index_trend("奇货贵金属")
    print(data)
    data = qhkc_index_profit_loss("奇货贵金属", end_date="20190716")
    print(data)
